consensus/event_loop_test.go

```diff
@@ -0,0 +1,899 @@
+			require.Fail(t, "Did not receive a block construction signal when there were enough timeouts to construct a timeout QC")
+		}
+
+		// Stop the event loop
+		fc.Stop()
+
+		// Confirm that the block construction signal has the expected parameters
+		require.Equal(t, signal.EventType, FastHotStuffEventTypeConstructTimeoutQC)
+		require.Equal(t, signal.View, uint64(5))                                           // The timeout QC will be proposed in view 5
+		require.Equal(t, signal.TipBlockHash.GetValue(), block1.GetBlockHash().GetValue()) // The timeout QC will be proposed in a block that extends from block 1
+		require.Equal(t, signal.TipBlockHeight, block1.GetHeight())                        // The timeout QC will be proposed at the block height after block 1
+		require.Equal(t, signal.AggregateQC.GetView(), uint64(4))                          // The timed out view is 4
+		require.Equal(t, signal.AggregateQC.GetHighQCViews(), []uint64{1, 2})              // The high QC view is 1 for validator 1 and 2 for validator 2
+		require.Equal(t,
+			signal.AggregateQC.GetAggregatedSignature().GetSignersList().ToBytes(),
+			bitset.NewBitset().Set(0, true).Set(1, true).ToBytes(), // Both validators have timed out, so both validators are in the timeout QC
+		)
+
+		// Verify that the high QC is the QC from block 2. It should be unchanged.
+		require.Equal(t, signal.AggregateQC.GetHighQC().GetBlockHash(), block2.GetQC().GetBlockHash())
+		require.Equal(t, signal.AggregateQC.GetHighQC().GetView(), block2.GetQC().GetView())
+		require.Equal(t,
+			signal.AggregateQC.GetHighQC().GetAggregatedSignature().GetSignersList().ToBytes(),
+			block2.GetQC().GetAggregatedSignature().GetSignersList().ToBytes(),
+		)
+		require.Equal(t,
+			signal.AggregateQC.GetHighQC().GetAggregatedSignature().GetSignature().ToBytes(),
+			block2.GetQC().GetAggregatedSignature().GetSignature().ToBytes(),
+		)
+	}
+}
+
+func TestFastHotStuffEventLoopStartStop(t *testing.T) {
+	oneHourInNanoSecs := time.Duration(3600000000000)
+
+	fc := NewFastHotStuffEventLoop()
+	err := fc.Init(oneHourInNanoSecs, 2*oneHourInNanoSecs,
+		BlockWithValidators{createDummyBlock(2), createDummyValidatorSet()},     // tip
+		[]BlockWithValidators{{createDummyBlock(2), createDummyValidatorSet()}}, // safeBlocks
+	)
+	require.NoError(t, err)
+
+	// Start the event loop
+	fc.Start()
+
+	// Confirm the event loop status has changed to running
+	require.Equal(t, eventLoopStatusRunning, fc.status)
+
+	// Confirm that the ETAs for the block construction and timeout timers have been set
+	require.Equal(t, fc.crankTimerTask.GetDuration(), oneHourInNanoSecs)
+	require.Equal(t, fc.nextTimeoutTask.GetDuration(), 2*oneHourInNanoSecs)
+
+	// Stop the event loop
+	fc.Stop()
+
+	// Confirm the event loop status has reverted to initialized
+	require.Equal(t, eventLoopStatusInitialized, fc.status)
+
+	// Confirm that calling fc.Stop() again doesn't panic
+	require.NotPanics(t, fc.Stop)
+}
```

Comment:
I skimmed these unit tests and I really like this. You start by testing that each function behaves correctly under all the edge-cases, then you can build up to the integration tests which I see in the next file.